### PR TITLE
[xh]Add OracleDB Destination

### PR DESCRIPTION
**Pull request by matrixstone**
# Tests
Tested with S3 source, destination is oracle DB.

Scenarios tested:
1. It can create table and export data.
2. If table already exists, it can export data directly.
3. If table schema is different, it can alter table
4. If insert data with unique keys or not and can update records if duplicated unique key records found with SQL/PL command.

Screenshot of the created table, exported records:
create table and generate output

Screenshot of the alter table schema:
alter_table

cc:
@wangxiaoyou1993 @dy46

### CODE REVIEW

**Comment by wangxiaoyou1993** [comment]
Could you add the README file?

File: mage_integrations/mage_integrations/destinations/oracledb/__init__.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+        table_name: str,
+        database_name: str = None,
+        unique_constraints: List[str] = None,
+    ) -> List[str]:
+        return [
+            build_create_table_command(
+                column_type_mapping=column_type_mapping(
+                    schema,
+                    convert_column_type,
+                    lambda item_type_converted: 'VARRAY(5)',
```
Comment (wangxiaoyou1993) [inline mage_integrations/mage_integrations/destinations/oracledb/__init__.py:87]: why is it `VARRAY(5)`? is 5 the maximum number of elements?

File: mage_integrations/mage_integrations/destinations/oracledb/__init__.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+                       not in current_columns]
+
+        if not new_columns:
+            return []
+        return [
+            build_alter_table_command(
+                column_type_mapping=column_type_mapping(
+                    schema,
+                    convert_column_type,
+                    lambda item_type_converted: 'TEXT',
```
Comment (wangxiaoyou1993) [inline mage_integrations/mage_integrations/destinations/oracledb/__init__.py:126]: this lambda function should be the one that's used in the `build_create_table_commands` method. this is used to map the array data type.
could you also test syncing array data type to oracle db?

File: mage_integrations/mage_integrations/destinations/oracledb/__init__.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        database_name: str = None,
+        unique_conflict_method: str = None,
+        unique_constraints: List[str] = None,
+    ) -> List[str]:
+        columns = list(schema['properties'].keys())
+        insert_columns, insert_values = build_insert_command(
+            column_type_mapping=column_type_mapping(
+                schema,
+                convert_column_type,
+                lambda item_type_converted: 'LONGTEXT',
```
Comment (wangxiaoyou1993) [inline mage_integrations/mage_integrations/destinations/oracledb/__init__.py:148]: again, this method should be the same with the one used in the `build_create_table_commands` and `build_alter_table_commands` methods.

File: mage_integrations/mage_integrations/destinations/oracledb/__init__.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+        self,
+        schema_name: str,
+        table_name: str,
+        database_name: str = None,
+    ) -> bool:
+        connection = self.build_connection().build_connection()
+        cursor = connection.cursor()
+        cursor.execute(
+            f'SELECT COUNT(*) FROM user_tables WHERE table_name = \'{table_name.upper()}\'')
+        (number_of_rows,) = cursor.fetchone()
```
Comment (wangxiaoyou1993) [inline mage_integrations/mage_integrations/destinations/oracledb/__init__.py:209]: you can just do `number_of_rows = cursor.fetchone()[0]`

File: mage_integrations/mage_integrations/destinations/oracledb/utils.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    COLUMN_FORMAT_DATETIME,
+    COLUMN_TYPE_BOOLEAN,
+    COLUMN_TYPE_INTEGER,
+    COLUMN_TYPE_NULL,
+    COLUMN_TYPE_NUMBER,
+    COLUMN_TYPE_OBJECT,
+    COLUMN_TYPE_STRING,
+)
+
+LOGGER = singer.get_logger()
```
Comment (wangxiaoyou1993) [inline mage_integrations/mage_integrations/destinations/oracledb/utils.py:19]: this is unused.

File: mage_integrations/mage_integrations/destinations/oracledb/__init__.py [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+        # While a "schema" is a user ID and a collection of the users tables and indexes.
+        return [
+            'SELECT name FROM v$database',
+        ]
+
+    def test_connection(self) -> None:
+        conn = self.build_connection().build_connection()
+        cursor = conn.cursor()
+        try:
+            cursor.execute("""SELECT name FROM v$database""")
```
Comment (wangxiaoyou1993) [inline mage_integrations/mage_integrations/destinations/oracledb/__init__.py:67]: If it's one line, we don't need three double quotes

File: mage_integrations/mage_integrations/destinations/oracledb/utils.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+
+def clean_column_name(col, lower_case: bool = True, handle_leading_underscore: bool = True):
```
Comment (wangxiaoyou1993) [inline mage_integrations/mage_integrations/destinations/oracledb/utils.py:22]: you should pass this arg to clean_column_name_orig as well.
also, if handle_leading_underscore is always false for oracledb, there's no need to add this optional arg to this method.

File: mage_integrations/mage_integrations/destinations/oracledb/__init__.py (hunk 6 above)
Comment (matrixstone) [reply mage_integrations/mage_integrations/destinations/oracledb/__init__.py:67]: Updated

File: mage_integrations/mage_integrations/destinations/oracledb/__init__.py (hunk 4 above)
Comment (matrixstone) [reply mage_integrations/mage_integrations/destinations/oracledb/__init__.py:209]: Updated

File: mage_integrations/mage_integrations/destinations/oracledb/utils.py (hunk 5 above)
Comment (matrixstone) [reply mage_integrations/mage_integrations/destinations/oracledb/utils.py:19]: Removed

File: mage_integrations/mage_integrations/destinations/oracledb/utils.py (hunk 7 above)
Comment (matrixstone) [reply mage_integrations/mage_integrations/destinations/oracledb/utils.py:20]: In alter command, handle_leading_underscore is set to be false, the rest of the time it is set to be default value as true

File: mage_integrations/mage_integrations/destinations/oracledb/__init__.py (hunk 1 above)
Comment (matrixstone) [reply mage_integrations/mage_integrations/destinations/oracledb/__init__.py]: Updated it to be LONG

File: mage_integrations/mage_integrations/destinations/oracledb/__init__.py (hunk 2 above)
Comment (matrixstone) [reply mage_integrations/mage_integrations/destinations/oracledb/__init__.py]: Updated it to be consistent with LONG type

File: mage_integrations/mage_integrations/destinations/oracledb/__init__.py (hunk 3 above)
Comment (matrixstone) [reply mage_integrations/mage_integrations/destinations/oracledb/__init__.py]: Updated it to be consistent with LONG type

File: mage_integrations/mage_integrations/destinations/oracledb/__init__.py [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+        table_name: str,
+        database_name: str = None,
+        unique_constraints: List[str] = None,
+    ) -> List[str]:
+        return [
+            build_create_table_command(
+                column_type_mapping=column_type_mapping(
+                    schema,
+                    convert_column_type,
+                    lambda item_type_converted: 'LONG',
```
Comment (wangxiaoyou1993) [inline mage_integrations/mage_integrations/destinations/oracledb/__init__.py:87]: I don't think it works. LONG is used to store number. This is lambda function is to map the array of item_type_converted to a column type.
Can you test syncing an array type?

**Comment by matrixstone** [comment]
Updated this PR and tested with Postgres source with Array type and Json type column.

For array type, the `varray` in oracle is a user defined collection type. Every time user needs to define such an array type `CREATE TYPE my_personal_type AS VARRAY(10) OF VARCHAR2(80);` with customized name `my_personal_type`. It shows `my_personal_type` when we read it. It is tricky to know the exact type. Currently I use a `CHAR(255)` type to convert array.

For json type, it is supported in this PR with customized `convert_column_to_type` function to convert to `NCLOB` type in Oracle.

Screenshots for the test table input and output:

int array input table:
test_array_input 3

int array output table:
test array output 1 3

Text array and Json column input table:
test array input2  3

Text array and Json column output table:
test array output 2 3

PTAL @wangxiaoyou1993

File: mage_integrations/mage_integrations/destinations/oracledb/__init__.py [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+        table_name: str,
+        database_name: str = None,
+        unique_constraints: List[str] = None,
+    ) -> List[str]:
+        return [
+            build_create_table_command(
+                column_type_mapping=column_type_mapping(
+                    schema,
+                    convert_column_type,
+                    lambda item_type_converted: 'CHAR(255)',
```
Comment (wangxiaoyou1993) [inline mage_integrations/mage_integrations/destinations/oracledb/__init__.py:88]: There could be limitation of CHAR(255) when inserting a large array. But I'll stamp first and merge this PR.
You can further test and address this issue in a separate PR.

**Review by wangxiaoyou1993 — APPROVED** [review]

**Pull request merged**